File: backend/app/routes/recall.py
"""POST /api/recall — semantic recall via pgvector."""

from __future__ import annotations

from fastapi import APIRouter, Request
from fastapi.responses import JSONResponse

from app.auth import CurrentUser
from app.gemini import embed_text
from app.supabase_client import get_user_client

router = APIRouter()


@router.post("/api/recall")
async def recall(request: Request, user: CurrentUser):
    try:
        body = await request.json()
    except Exception:
        return JSONResponse({"error": "Invalid JSON body"}, status_code=400)

    session_id = body.get("sessionId")
    query = (body.get("query") or "").strip()
    match_count = body.get("matchCount", 5)

    if not session_id or not query:
        return JSONResponse({"error": "sessionId and query are required"}, status_code=400)

    token = user.get("_access_token", "")
    supabase = get_user_client(token)
    if not supabase:
        return {"hits": [], "context": ""}

    session = (
        supabase.table("chat_sessions")
        .select("id")
        .eq("id", session_id)
        .eq("user_id", user["id"])
        .maybe_single()
        .execute()
    )
    if not session.data:
        return {"hits": [], "context": ""}

    try:
        embedding = await embed_text(query)
    except Exception:
        embedding = None
    if not embedding:
        return {"hits": [], "context": ""}

    try:
        res = supabase.rpc(
            "match_chat_embeddings",
            {
                "p_session_id": session_id,
                "p_query_embedding": embedding,
                "p_match_count": match_count,
            },
        ).execute()
        hits = res.data or []
    except Exception:
        return {"hits": [], "context": ""}

    context = ""
    if hits:
        lines = [f"- ({h.get('role')}) {(h.get('content') or '')[:300]}" for h in hits]
        context = "[Relevant context from earlier in this chat]\n" + "\n".join(lines)

    return {"hits": hits, "context": context}

File: lib/memory-store/recall.ts
import type { SupabaseClient } from '@supabase/supabase-js';
import { embedText } from '@/lib/embeddings';
import type { RecallHit, RecallResult } from './types';

export interface RecallSimilarTurnsInput {
  sessionId: string;
  query: string;
  matchCount?: number;
  userId?: string;
}

function buildRecallContextBlock(hits: RecallHit[]): string {
  if (!hits.length) return '';
  return `[Relevant context from earlier in this chat]\n${hits
    .map(h => `- (${h.role}) ${h.content.slice(0, 300)}`)
    .join('\n')}`;
}

export async function recallSimilarTurns(
  supabase: SupabaseClient,
  input: RecallSimilarTurnsInput,
): Promise<RecallResult> {
  const { sessionId, query, matchCount = 5, userId } = input;

  if (!sessionId || !query?.trim()) {
    return { hits: [], contextBlock: '' };
  }

  if (userId) {
    const { data: session } = await supabase
      .from('chat_sessions')
      .select('id')
      .eq('id', sessionId)
      .eq('user_id', userId)
      .maybeSingle();
    if (!session) {
      return { hits: [], contextBlock: '' };
    }
  }

  const embedding = await embedText(query);
  if (!embedding) {
    return { hits: [], contextBlock: '' };
  }

  const { data, error } = await supabase.rpc('match_chat_embeddings', {
    p_session_id: sessionId,
    p_query_embedding: embedding as unknown as string,
    p_match_count: matchCount,
  });

  if (error) {
    console.error('[recall rpc]', error.message);
    return { hits: [], contextBlock: '' };
  }

  const hits = (data ?? []) as RecallHit[];
  return {
    hits,
    contextBlock: buildRecallContextBlock(hits),
  };
}
